enable quicker tcp error notification

engelmi opened the pull request:
Fixes: https://github.com/eclipse-bluechi/bluechi/issues/652
Relates to: https://github.com/eclipse-bluechi/bluechi/issues/648

When a connection between controller and agent is esablished and is dropped later, e.g. removing the cable from the agent, the disconnect is properly detected by the keepalive mechanism. However, the keepalive takes a while to detect this (based on the KEEPCNT of the system). If any command is issued during that time frame, tcp will try to retransmit the data. Eventually, retransmitting will be stopped and ICMP packets for host not reachable emitted. The keepalive is not used then, which results in the connection between agent and controller to be broken but not closed - a disconnect is not detected.
By setting the IP_RECVERR socket option, errors such as the ICMP host not reachable error will be delivered to the upper layer (systemd event loop in our case) so that it can handle it.

A detailed explanation can be found in this comment: https://github.com/eclipse-bluechi/bluechi/issues/652#issuecomment-1859772124

engelmi committed:
enable quicker tcp error notification

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/652
Relates to: https://github.com/eclipse-bluechi/bluechi/issues/648

When a connection between controller and agent is esablished and
is dropped later, e.g. removing the cable from the agent, the
disconnect is properly detected by the keepalive mechanism.
However, the keepalive takes a while to detect this (based on the
KEEPCNT of the system). If any command is issued during that time
frame, tcp will try to retransmit the data. Eventually, retransmitting
will be stopped and ICMP packets for host not reachable emitted.
The keepalive is not used then, which results in the connection
between agent and controller to be broken but not closed - a
disconnect is not detected.
By setting the IP_RECVERR socket option, errors such as the ICMP
host not reachable error will be delivered to the upper layer
(systemd event loop in our case) so that it can handle it.

Signed-off-by: Michael Engel <mengel@redhat.com>

mkemel reviewed (approved):
Impressive debugging work

engelmi committed:
made IP_RECV_ERR configurable

Relates to: https://github.com/eclipse-bluechi/bluechi/issues/652

For systems with less robust network, always enabling IP_RECV_ERR
might result in unnecessary disconnects and a less robust BlueChi
setup. Therefore, a configuration setting was introduced to enable
or disable this option depending on the current needs.

Signed-off-by: Michael Engel <mengel@redhat.com>

alexlarsson commented on src/agent/agent.c:
```diff
@@ -623,6 +624,8 @@ bool agent_parse_config(Agent *agent, const char *configfile) {
                 }
         }
 
+        agent->ip_receive_errors = cfg_get_bool_value(agent->config, CFG_IP_RECEIVE_ERRORS);
```
This seems to default to false, but the docs said true?

alexlarsson replied:
Ah, no I see. we actually force set it in the config object, so it will always be set.

engelmi replied:
Yes, that is due to how we read bool values from a config file - using a list of values mapped to true, everything else is false. 
But we set all default values right before reading config files in `libbluechi/common/cfg.c`, e.g. via `cfg_def_conf(...)`. So its guaranteed that the config struct contains at least the default. The default is defined in `libbluechi/common/protocol.h`.
